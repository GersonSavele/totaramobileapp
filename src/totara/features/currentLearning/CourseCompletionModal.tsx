/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import React from "react";
import { translate } from "@totara/locale";
import { PrimaryButton, InfoModal } from "@totara/components";
import { Images } from "@resources/images";
import { ImageSourcePropType } from "react-native";

type CourseCompletionProps = {
  onClose: () => void;
};

const CourseCompletionModal = ({ onClose }: CourseCompletionProps) => {
  return (
    <InfoModal
      title={translate("course.course_complete.title")}
      description={translate("course.course_complete.description")}
      imageSource={Images.courseComplete as ImageSourcePropType}
      visible>
      <PrimaryButton text={translate("course.course_complete.button_title")} onPress={onClose} />
    </InfoModal>
  );
};

export default CourseCompletionModal;
