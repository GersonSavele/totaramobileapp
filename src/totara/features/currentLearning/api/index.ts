/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import { gql } from "@apollo/client";
import { config } from "@totara/lib";

const query = config.mobileApi.persistentQuery
  ? gql`
      query totara_mobile_current_learning {
        currentLearning: totara_mobile_current_learning {
          id
          itemtype
          itemcomponent
          shortname
          fullname
          summary: description(format: PLAIN)
          progress
          urlView: url_view
          duedate(format: ISO8601)
          duedateState: duedate_state
          native: mobile_coursecompat
          imageSrc: mobile_image
          summaryFormat: description_format
          __typename
        }
      }
    `
  : gql`
      query mobile_currentlearning_my_items {
        currentLearning: mobile_currentlearning_my_items {
          id
          itemtype
          itemcomponent
          shortname
          fullname
          summary: description(format: PLAIN)
          progress
          urlView: url_view
          duedate(format: ISO8601)
          duedateState: duedate_state
          native: mobile_coursecompat
          imageSrc: mobile_image
          summaryFormat: description_format
          __typename
        }
      }
    `;

export default query;
