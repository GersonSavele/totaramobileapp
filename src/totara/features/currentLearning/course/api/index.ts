/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import { gql } from "@apollo/client";
import { config } from "@totara/lib";
import { AUTH_HEADER_FIELD } from "@totara/lib/constants";

const coreCourse = gql`
  query totara_mobile_course($courseid: core_id!, $guestpw: String) {
    mobile_course: totara_mobile_course(courseid: $courseid, guestpw: $guestpw) {
      course(courseid: $courseid) {
        id
        format
        fullname
        shortname
        summary(format: PLAIN)
        summaryformat
        startdate(format: ISO8601)
        enddate(format: ISO8601)
        lang
        image
        sections {
          id
          title
          available
          availablereason(format: PLAIN)
          summary(format: PLAIN)
          summaryformat
          data: modules {
            id
            instanceid
            modtype
            name
            available
            availablereason(format: PLAIN)
            viewurl
            completion
            completionstatus
            showdescription
            description(format: PLAIN)
            gradefinal
            gradepercentage
            descriptionformat
            __typename
          }
          __typename
        }
        criteriaaggregation
        criteria {
          id
          type
          typeaggregation
          criteria
          requirement
          status
          complete
          completiondate(format: ISO8601)
          __typename
        }
        showGrades: showgrades
        completionEnabled: completionenabled
        completion {
          id
          statuskey
          progress
          timecompleted(format: ISO8601)
          __typename
        }
        __typename
      }
      native: mobile_coursecompat
      imageSrc: mobile_image
      gradeFinal: formatted_gradefinal
      gradeMax: formatted_grademax
      __typename
    }
  }
`;

const courseSelfComplete = gql`
  mutation totara_mobile_completion_course_self_complete($courseid: core_id!) {
    core_completion_course_self_complete(courseid: $courseid)
  }
`;

const activitySelfComplete = gql`
  mutation totara_mobile_completion_activity_self_complete($cmid: core_id!, $complete: Boolean!) {
    core_completion_activity_self_complete(cmid: $cmid, complete: $complete)
  }
`;

type FetchParam = {
  instanceId: number;
  apiKey: string;
  host: string;
  modtype?: string;
};

const XMLHttpRequestPostFetch = (host, body, apiKey) => {
  return new Promise(function (resolve, reject) {
    // We decided to use XMLHttpRequest instead of fetch because fetch seems to have an issue being blocked by some missing certificate
    // please check https://github.com/facebook/react-native/issues/32931 and MOB-1144
    // eslint-disable-next-line no-undef
    var request = new XMLHttpRequest();
    request.open("POST", config.apiUri(host));
    request.setRequestHeader("Content-Type", "application/json;charset=UTF-8");
    request.setRequestHeader(AUTH_HEADER_FIELD, apiKey);

    request.onload = function () {
      if (request.status >= 200 && request.status < 300) {
        resolve(request.response);
      } else {
        reject({
          status: request.status,
          statusText: request.statusText
        });
      }
    };

    request.onerror = function () {
      reject({
        status: request.status,
        statusText: request.statusText
      });
    };

    request.send(body);
  });
};

const fetchResource = ({ instanceId, apiKey, host }: FetchParam): Promise<unknown> => {
  return XMLHttpRequestPostFetch(
    host,
    JSON.stringify({
      operationName: "totara_mobile_resource",
      variables: { resourceid: instanceId }
    }),
    apiKey
  );
};

const updateStateViewResource = ({ instanceId, modtype, apiKey, host }: FetchParam): Promise<unknown> => {
  return XMLHttpRequestPostFetch(
    host,
    JSON.stringify({
      operationName: "totara_mobile_completion_activity_view",
      variables: { cmid: instanceId, activity: modtype }
    }),
    apiKey
  );
};

export { coreCourse, courseSelfComplete, activitySelfComplete, fetchResource, updateStateViewResource };
