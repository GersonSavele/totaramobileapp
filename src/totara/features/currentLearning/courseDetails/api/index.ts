/**
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Jun Yamog <jun.yamog@totaralearning.com>
 *
 */

import gql from "graphql-tag";

const coreCourse = gql`
  query totara_mobile_course($courseid: ID!) {
    mobile_course: totara_mobile_course(courseid: $courseid) {
      course(courseid: $courseid) {
        id
        fullname
        shortname
        summary(format: PLAIN)
        startdate(format: ISO8601)
        enddate(format: ISO8601)
        lang
        image
        sections {
          id
          title
          available
          availablereason(format: PLAIN)
          data: modules {
            id
            instanceid
            modtype
            name
            available
            availablereason(format: PLAIN)
            viewurl
            completion
            completionstatus
            showdescription
            description(format: PLAIN)
            gradefinal
            gradepercentage
            __typename
          }
          __typename
        }
        criteriaaggregation
        criteria {
          id
          type
          typeaggregation
          criteria
          requirement
          status
          complete
          completiondate(format: ISO8601)
          __typename
        }
        completion {
          id
          statuskey
          progress
          timecompleted(format: ISO8601)
          gradefinal
          grademax
          __typename
        }
        __typename
      }
      native: mobile_coursecompat
      imageSrc: mobile_image
      __typename
    }
  }
`;

const courseSelfComplete = gql`
  mutation totara_mobile_completion_course_self_complete($courseid: core_id!) {
    core_completion_course_self_complete(courseid: $courseid)
  }
`;

export { coreCourse, courseSelfComplete };
