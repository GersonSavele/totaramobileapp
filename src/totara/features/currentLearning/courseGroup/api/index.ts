/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import { gql } from "@apollo/client";

const coreProgram = gql`
  query totara_mobile_program($programid: core_id!) {
    totara_mobile_program(programid: $programid) {
      id
      fullname
      shortname
      duedate(format: ISO8601)
      duedateState: duedate_state(format: PLAIN)
      summary(format: PLAIN)
      summaryformat
      endnote(format: PLAIN)
      availablefrom(format: ISO8601)
      availableuntil(format: ISO8601)
      imageSrc: mobile_image
      completion {
        id
        statuskey
        progress
        __typename
      }
      currentCourseSets: current_coursesets {
        id
        label
        nextsetoperator
        completionCriteria: criteria
        courses {
          id
          itemtype
          itemcomponent
          shortname
          fullname
          summary: description(format: PLAIN)
          summaryFormat
          progress
          urlView: url_view
          duedate(format: ISO8601)
          duedateState: duedate_state
          native: mobile_coursecompat
          imageSrc: mobile_image
          viewable
          __typename
        }
        __typename
      }
      countUnavailableSets: count_unavailablesets
      courseSetHeader: courseset_header
      __typename
    }
  }
`;

const coreCertification = gql`
  query totara_mobile_certification($certificationid: core_id!) {
    totara_mobile_certification(certificationid: $certificationid) {
      id
      fullname
      shortname
      duedate(format: ISO8601)
      duedateState: duedate_state(format: PLAIN)
      summary(format: PLAIN)
      summaryformat
      endnote(format: PLAIN)
      availablefrom(format: ISO8601)
      availableuntil(format: ISO8601)
      imageSrc: mobile_image
      completion {
        id
        statuskey
        renewalstatuskey
        progress
        __typename
      }
      currentCourseSets: current_coursesets {
        id
        label
        nextsetoperator
        completionCriteria: criteria
        courses {
          id
          itemtype
          itemcomponent
          shortname
          fullname
          summary: description(format: PLAIN)
          summaryFormat
          progress
          urlView: url_view
          duedate(format: ISO8601)
          duedateState: duedate_state
          native: mobile_coursecompat
          imageSrc: mobile_image
          viewable
          __typename
        }
        __typename
      }
      countUnavailableSets: count_unavailablesets
      courseSetHeader: courseset_header
      __typename
    }
  }
`;

const mutationReportProgramme = gql`
  mutation totara_mobile_program_view($program_id: core_id!) {
    totara_mobile_program_view(program_id: $program_id)
  }
`;

export { coreProgram, coreCertification, mutationReportProgramme };
