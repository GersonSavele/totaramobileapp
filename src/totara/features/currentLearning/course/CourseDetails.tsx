/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTDâ€™s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import React, { useState } from "react";
import { StyleSheet, View, Text, Switch } from "react-native";
import { isEmpty, isEqual } from "lodash";
import { Loading, LoadingError } from "@totara/components";
import { useQuery, NetworkStatus } from "@apollo/client";
import { translate } from "@totara/locale";
import { coreCourse } from "./api";
import Activities from "./Activities";
import { CourseContentDetails } from "@totara/types";
import { StatusKey } from "@totara/types/Completion";
import OverviewDetails from "../overview/OverviewDetails";
import { TotaraTheme } from "@totara/theme/Theme";
import LearningDetails from "../LearningDetails";
import CourseCompletionModal from "../CourseCompletionModal";
import { learningItemEnum } from "@totara/features/constants";
import courseDetailsStyle from "./courseDetailsStyle";
import { CourseFormat } from "@totara/types/Course";
import { DescriptionFormat } from "@totara/types/LearningItem";

const CourseDetails = ({ navigation, route }: any) => {
  const courseId = route?.params?.targetId || navigation.getParam("targetId");
  const passwordRequired = route?.params?.passwordRequired;

  let queryVariables = { courseid: courseId };
  if (passwordRequired) {
    const guestPassword = route?.params?.guestPassword;
    queryVariables = { ...queryVariables, guestpw: guestPassword };
  }

  const { networkStatus, error, data, refetch } = useQuery(coreCourse, {
    variables: queryVariables,
    notifyOnNetworkStatusChange: true
  });

  const onContentRefresh = () => {
    refetch();
  };

  if (networkStatus === NetworkStatus.loading) return <Loading />;
  if (!data && error) return <LoadingError onRefreshTap={onContentRefresh} error={error} />;

  if (data) {
    return (
      <CourseDetailsContent
        loading={networkStatus === NetworkStatus.refetch}
        pullToRefresh={onContentRefresh}
        courseDetails={data.mobile_course}
        courseRefreshCallback={refetch}
        navigation={navigation}
      />
    );
  } else {
    return <Loading />;
  }
};

type CourseDetailsContentProps = {
  courseDetails: CourseContentDetails;
  courseRefreshCallback: () => {};
  pullToRefresh: () => void;
  navigation: any;
  loading: boolean;
};

const CourseDetailsContent = ({
  courseDetails,
  courseRefreshCallback,
  pullToRefresh,
  navigation,
  loading
}: CourseDetailsContentProps) => {
  const expanableSections =
    courseDetails?.course?.sections.filter(section => section.available && !isEmpty(section.data)) || [];
  const expanableSectionIds = Array.from(expanableSections, section => section.id);
  const isSingleActivity = courseDetails?.course?.format === CourseFormat.singleActivity;

  const [showOverview, setShowOverview] = useState(true);
  const [showCompletionModal, setShowCompletionModal] = useState(true);
  const [expandedSectionIds, setExpandedSectionIds] = useState<number[]>(isSingleActivity ? expanableSectionIds : []);

  const isExpandedAll = !isEmpty(expanableSectionIds) && isEqual(expanableSectionIds.sort(), expandedSectionIds.sort());

  const onClose = () => {
    setShowCompletionModal(!showCompletionModal);
    courseRefreshCallback();
    navigation.goBack();
  };

  const onSwitchTab = () => {
    setShowOverview(!showOverview);
  };

  const onChangeExpand = (isExpanded: boolean, sectionIds: number[]) => {
    if (isExpanded) {
      setExpandedSectionIds([]);
    } else {
      setExpandedSectionIds(sectionIds);
    }
  };

  return (
    <LearningDetails
      onPullToRefresh={pullToRefresh}
      loading={loading}
      item={courseDetails.course}
      itemType={learningItemEnum.Course}
      tabBarLeftTitle={translate("course.course_details.overview")}
      tabBarRightTitle={translate("course.course_details.activities")}
      onPress={onSwitchTab}
      overviewIsShown={showOverview}
      image={courseDetails.imageSrc}
      badgeTitle={translate("learning_items.course")}
      navigation={navigation}>
      <View style={[styles.container, { backgroundColor: TotaraTheme.colorNeutral2 }]}>
        <View style={[styles.activitiesContainer, { backgroundColor: TotaraTheme.colorNeutral1 }]}>
          {!showOverview ? (
            <View
              style={{
                backgroundColor: TotaraTheme.colorAccent
              }}>
              {!isSingleActivity && (
                <View style={courseDetailsStyle.expandContentWrap}>
                  <Text style={courseDetailsStyle.expandTextWrap}>
                    {translate("course.course_details.expand_or_collapse")}
                  </Text>
                  <Switch
                    style={[{ borderColor: TotaraTheme.colorNeutral5 }]}
                    value={isExpandedAll}
                    onValueChange={() => onChangeExpand(isExpandedAll, expanableSectionIds)}
                    accessibilityLabel={translate("course.course_details.accessibility_expand_all")}
                  />
                </View>
              )}
              <Activities
                sections={courseDetails.course.sections}
                courseRefreshCallBack={courseRefreshCallback}
                expandedSectionIds={expandedSectionIds}
                onSetExpandedSectionIds={setExpandedSectionIds}
                completionEnabled={courseDetails.course.completionEnabled}
                isSingleActivity={isSingleActivity}
              />
            </View>
          ) : (
            <OverviewDetails
              isCourseSet={false}
              id={courseDetails.course.id}
              criteria={courseDetails.course.criteria}
              summary={courseDetails.course.summary}
              summaryFormat={courseDetails.course.summaryformat as DescriptionFormat}
              gradeFinal={courseDetails.gradeFinal}
              progress={courseDetails.course.completion.progress}
              summaryTypeTitle={translate("course.course_overview.course_summary")}
              onclickContinueLearning={onClose}
              courseRefreshCallback={courseRefreshCallback}
              showGrades={courseDetails.course.showGrades}
              completionEnabled={courseDetails.course.completionEnabled}
            />
          )}
        </View>
      </View>
      {courseDetails.course.completion && courseDetails.course.completion.statuskey === StatusKey.complete && (
        <CourseCompletionModal onClose={onClose} />
      )}
    </LearningDetails>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  activitiesContainer: {
    flex: 3,
    padding: 0
  }
});

export default CourseDetails;
