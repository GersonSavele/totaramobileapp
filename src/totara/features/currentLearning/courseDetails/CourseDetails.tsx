/**
 * This file is part of Totara Mobile
 *
 * Copyright (C) 2019 onwards Totara Learning Solutions LTD
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Jun Yamog <jun.yamog@totaralearning.com>
 */

import React, { useState, useContext } from "react";
import { StyleSheet, View } from "react-native";
import {
  withNavigation,
  NavigationParams,
  NavigationInjectedProps
} from "react-navigation";
import { GeneralErrorModal } from "@totara/components";
import { useQuery } from "@apollo/react-hooks";
import { translate } from "@totara/locale";
import { coreCourse } from "./api";
import Activities from "./Activities";
import { CourseContentDetails } from "@totara/types";
import OverviewDetails from "../overview/Details";
import { ThemeContext } from "@totara/theme";
import {
  HeaderView,
  CourseCompletionModal
} from "@totara/components/currentLearning";
import ActivitySheetWrapper from "@totara/activities/ActivitySheetWrapper";
import { StatusKey } from "@totara/lib/constants";

type CourseDetailsProps = {
  courseDetails: CourseContentDetails;
  refetch: () => {};
  navigation?: NavigationParams;
};

const CourseDetails = ({ navigation }: NavigationInjectedProps) => {
  const courseId = navigation.getParam("targetId");
  const { loading, error, data, refetch } = useQuery(coreCourse, {
    variables: { courseid: courseId }
  });
  if (loading) return null;
  if (error) return <GeneralErrorModal siteUrl="" />;
  if (data) {
    return (
      <ActivitySheetWrapper>
        <CourseDetailsComponent
          courseDetails={data.mobile_course}
          refetch={refetch}
        />
      </ActivitySheetWrapper>
    );
  }
};

const CourseDetailsComponent = withNavigation(
  ({ navigation = {}, courseDetails, refetch }: CourseDetailsProps) => {
    const [showOverview, setShowOverview] = useState(true);
    const [showCompletionModal, setShowCompletionModal] = useState(true);

    const onClose = () => {
      setShowCompletionModal(!showCompletionModal);
      navigation.goBack();
    };
    const onSwitchTab = () => {
      setShowOverview(!showOverview);
    };
    const [theme] = useContext(ThemeContext);
    return (
      <HeaderView
        details={courseDetails.course}
        navigation={navigation!}
        tabBarLeft={translate("course.course-details.overview")}
        tabBarRight={translate("course.course-details.activities")}
        onPress={onSwitchTab}
        showOverview={showOverview}
        image={courseDetails.imageSrc}
        badgeTitle="Course">
        <View
          style={[styles.container, { backgroundColor: theme.colorNeutral2 }]}>
          <View
            style={[
              styles.activitiesContainer,
              { backgroundColor: theme.colorNeutral1 }
            ]}>
            {!showOverview ? (
              <Activities
                sections={courseDetails.course.sections}
                refetch={refetch}
              />
            ) : (
              <OverviewDetails
                learningItem={courseDetails.course}
                summaryTypeTitle={translate(
                  "course.course_overview.course_summery"
                )}
                onclickContinueLearning={onClose}
              />
            )}
          </View>
        </View>
        {courseDetails.course.completion &&
          courseDetails.course.completion.statuskey === StatusKey.complete && (
            <CourseCompletionModal onClose={onClose} />
          )}
      </HeaderView>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  activitiesContainer: {
    flex: 3,
    padding: 0
  }
});

export default CourseDetails;
