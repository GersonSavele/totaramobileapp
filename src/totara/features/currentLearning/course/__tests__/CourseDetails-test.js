/**
 * This file is part of Totara Enterprise.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise is provided only to Totara Learning Solutions
 * LTD’s customers and partners, pursuant to the terms and
 * conditions of a separate agreement with Totara Learning
 * Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [sales@totaralearning.com] for more information.
 */

import { MockedProvider } from '@apollo/client/testing';
import TestRenderer from "react-test-renderer";
import React from "react";
import wait from "waait";

import { coreCourse } from "../api";
import CourseDetails from "../CourseDetails";
import CourseCompletionModal from "../../CourseCompletionModal";
import { act } from "@testing-library/react-native";
const response = {
  course: course
};

const course = {
  completion: {
    gradefinal: 0,
    grademax: 100,
    id: 63,
    progress: 16,
    statuskey: "complete",
    timecompleted: null
  },
  criteria: [
    {
      complete: false,
      completiondate: null,
      criteria: "<a href=https://mobile.demo.totara.software/mod/scorm/view.php?id=398>Creating a set audience</a>",
      id: "106",
      requirement: "Achieving grade, Completed",
      status: "Has not achieved grade",
      type: "Activity completion",
      typeaggregation: "Any"
    }
  ],

  criteriaaggregation: "All",
  enddate: null,
  fullname: "(BETA) Audiences in Totara",
  id: "4",
  imageSrc: "https://mobile.demo.totara.software/totara/mobile/pluginfile.php/688/course/images/1588546858/image",
  lang: "",
  sections: [
    {
      available: true,
      availablereason: "",
      data: [
        {
          available: true,
          availablereason: "",
          completion: "tracking_manual",
          completionstatus: "complete",
          description:
            "Welcome to the Audiences forum. This area is for discussions on the course topic as well as questions about the course work_.",
          gradefinal: null,
          gradepercentage: 0,
          id: "371",
          instanceid: "22",
          modtype: "forum",
          name: "Course forum",
          showdescription: false,
          viewurl: "https://mobile.demo.totara.software/mod/forum/view.php?id=371"
        }
      ],
      id: "174",
      title: "Ask Us"
    }
  ],
  shortname: "(BETA) Audiences",
  startdate: "2017-08-24T00:00:00+0100",
  summary:
    "GROUPING YOUR USERS TO PROVIDE A PERSONALISED LEARNING EXPERIENCE↵↵Audiences are a powerful tool in Totara Learn, allowing you to group your users in order to assign them learning and performance management activities.↵↵Enrol in this course to explore how to:↵↵	* Create set and dynamic audiences↵↵	* Assign learning to an audience ↵↵ The course will take you around one hour 15 minutes to complete.↵↵"
};

const navigation = { getParam: jest.fn() };
const mocks = [
  {
    request: {
      query: coreCourse,
      variables: { courseId: 4 }
    },
    result: {
      data: response
    }
  }
];

const mocksError = [
  {
    request: {
      query: coreCourse,
      variables: { courseId: 8 }
    },
    error: new Error("Error")
  }
];

describe("Testing: Apollo MockedProvider should test three state such as loading, final and error", () => {
  it("Test result : Render loading state initially and return no component", async () => {
    let component;
    await TestRenderer.act(async () => {
      component = TestRenderer.create(
        <MockedProvider mocks={[]} addTypename={false}>
          <CourseDetails navigation={navigation} />
        </MockedProvider>
      );
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Test result : Once render, it will return child component", async () => {
    let component;
    await TestRenderer.act(async () => {
      component = TestRenderer.create(
        <MockedProvider mocks={mocks} addTypename={false}>
          <CourseDetails navigation={navigation} />
        </MockedProvider>
      );
    });
    await act(async () => {
      await wait(0);
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Test result : Set mock-error response and return Error UI", async () => {
    let component;
    await TestRenderer.act(async () => {
      component = TestRenderer.create(
        <MockedProvider mocks={mocksError} addTypename={false}>
          <CourseDetails navigation={navigation} />
        </MockedProvider>
      );
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Testing: Course manual completion", () => {
  it("Test result : course complete and loading action modal to confirm", async () => {
    let component;
    await TestRenderer.act(async () => {
      component = TestRenderer.create(<CourseCompletionModal onClose={jest.fn()} />);
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
});
