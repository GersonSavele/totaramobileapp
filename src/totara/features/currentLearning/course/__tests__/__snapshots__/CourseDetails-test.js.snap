// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing: Apollo MockedProvider should test three state such as loading, final and error Test result : Once render, it will return child component 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "column",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignContent": "space-between",
        "alignItems": "center",
        "padding": 8,
      }
    }
  >
    <Image
      source="test-file-stub"
    />
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 32,
          "fontWeight": "700",
          "lineHeight": 41,
          "padding": 8,
          "textAlign": "center",
        }
      }
    >
      translated content_error.title with undefined
    </Text>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 17,
          "fontWeight": "400",
          "lineHeight": 22,
          "marginHorizontal": 16,
          "textAlign": "center",
        }
      }
    >
      translated content_error.description with undefined
    </Text>
  </View>
  <View
    style={
      Object {
        "padding": 8,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          disabled={false}
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          this is a mock button
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Testing: Apollo MockedProvider should test three state such as loading, final and error Test result : Render loading state initially and return no component 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <Text>
    mock spinner
  </Text>
</View>
`;

exports[`Testing: Apollo MockedProvider should test three state such as loading, final and error Test result : Set mock-error response and return Error UI 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <Text>
    mock spinner
  </Text>
</View>
`;

exports[`Testing: Course manual completion Test result : course complete and loading action modal to confirm 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.3)",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 4,
          "justifyContent": "center",
          "marginHorizontal": 24,
          "paddingVertical": 32,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "marginBottom": 16,
            "marginHorizontal": 24,
          }
        }
      >
        <Image
          source="test-file-stub"
          style={
            Object {
              "alignItems": "center",
              "resizeMode": "contain",
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "marginBottom": 16,
            "marginHorizontal": 24,
          }
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={2}
          style={
            Object {
              "color": "#3D444B",
              "fontSize": 32,
              "fontWeight": "700",
              "lineHeight": 41,
              "paddingBottom": 16,
              "textAlign": "center",
            }
          }
        >
          translated course.course_complete.title with undefined
        </Text>
        <Text
          ellipsizeMode="tail"
          numberOfLines={3}
          style={
            Object {
              "color": "#3D444B",
              "fontSize": 17,
              "fontWeight": "400",
              "lineHeight": 22,
              "paddingBottom": 16,
              "textAlign": "center",
            }
          }
        >
          translated course.course_complete.description with undefined
        </Text>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
              ]
            }
          >
            <Text
              disabled={false}
              style={
                Array [
                  Object {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              this is a mock button
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
